core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/sync/StackSyncFinishedAction.java

```diff
@@ -0,0 +1,288 @@
+
+import static com.sequenceiq.cloudbreak.common.type.Status.AVAILABLE;
+import static com.sequenceiq.cloudbreak.common.type.Status.CREATE_FAILED;
+import static com.sequenceiq.cloudbreak.common.type.Status.DELETE_FAILED;
+import static com.sequenceiq.cloudbreak.common.type.Status.STOPPED;
+import static com.sequenceiq.cloudbreak.common.type.Status.WAIT_FOR_SYNC;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.sequenceiq.cloudbreak.cloud.event.resource.GetInstancesStateResult;
+import com.sequenceiq.cloudbreak.cloud.model.CloudVmInstanceStatus;
+import com.sequenceiq.cloudbreak.common.type.HostMetadataState;
+import com.sequenceiq.cloudbreak.common.type.InstanceStatus;
+import com.sequenceiq.cloudbreak.common.type.Status;
+import com.sequenceiq.cloudbreak.controller.NotFoundException;
+import com.sequenceiq.cloudbreak.domain.Cluster;
+import com.sequenceiq.cloudbreak.domain.HostMetadata;
+import com.sequenceiq.cloudbreak.domain.InstanceGroup;
+import com.sequenceiq.cloudbreak.domain.InstanceMetaData;
+import com.sequenceiq.cloudbreak.domain.Resource;
+import com.sequenceiq.cloudbreak.domain.Stack;
+import com.sequenceiq.cloudbreak.repository.HostMetadataRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceGroupRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceMetaDataRepository;
+import com.sequenceiq.cloudbreak.repository.ResourceRepository;
+import com.sequenceiq.cloudbreak.repository.StackUpdater;
+import com.sequenceiq.cloudbreak.service.cluster.flow.AmbariClusterConnector;
+import com.sequenceiq.cloudbreak.service.events.CloudbreakEventService;
+import com.sequenceiq.cloudbreak.service.messages.CloudbreakMessagesService;
+import com.sequenceiq.cloudbreak.service.stack.flow.InstanceSyncState;
+
+@Component("StackSyncFinishedAction")
+public class StackSyncFinishedAction extends AbstractStackSyncAction<GetInstancesStateResult> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackSyncFinishedAction.class);
+
+    @Inject
+    private InstanceGroupRepository instanceGroupRepository;
+    @Inject
+    private HostMetadataRepository hostMetadataRepository;
+    @Inject
+    private ResourceRepository resourceRepository;
+    @Inject
+    private AmbariClusterConnector ambariClusterConnector;
+    @Inject
+    private StackUpdater stackUpdater;
+    @Inject
+    private CloudbreakEventService eventService;
+    @Inject
+    private CloudbreakMessagesService cloudbreakMessagesService;
+    @Inject
+    private InstanceMetaDataRepository instanceMetaDataRepository;
+
+    public StackSyncFinishedAction() {
+        super(GetInstancesStateResult.class);
+    }
+
+    @Override
```

Comment:
I would create a new method in com.sequenceiq.cloudbreak.service.stack.flow.StackSyncService with the list of instance states and move the business logic there